tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -4246,6 +4246,52 @@ void CGMSHLSLRuntime::SetPatchConstantFunctionWithAttr(
   
 }
 
+static GlobalVariable *CreateStaticGlobal(llvm::Module *M, GlobalVariable *GV) {
+  Constant *GC = M->getOrInsertGlobal(GV->getName().str() + "_static_copy",
+                                      GV->getType()->getPointerElementType());
+  GlobalVariable *NGV = cast<GlobalVariable>(GC);
+  if (GV->hasInitializer()) {
+    NGV->setInitializer(GV->getInitializer());
+  }
+  // static global should have internal linkage
+  NGV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+  return NGV;
+}
+
+static bool CreateWriteEnabledStaticGlobals(llvm::Module *M) {
+  std::vector<GlobalVariable *> worklist;
+  for (GlobalVariable &GV : M->globals()) {
+    if (!GV.isConstant()) {
+      worklist.emplace_back(&GV);
+      GV.setConstant(true);
+    }
+  }
+
+  for (GlobalVariable *GV : worklist) {
+    std::set<BasicBlock *> entryBlocks;
+    for (User *U : GV->users()) {
+      if (Instruction *I = dyn_cast<Instruction>(U)) {
+        BasicBlock *entryBlock =
+            &(I->getParent()->getParent()->getEntryBlock());
+        if (entryBlocks.find(entryBlock) == entryBlocks.end())
+          entryBlocks.insert(entryBlock);
+      }
+    }
+
+    GlobalVariable *NGV = CreateStaticGlobal(M, GV);
+    // insert memcpy in all entryblocks
```

Comment:
You only want to insert memcpy into entry block of main EntryFunction, not all functions.

Reply:
fixed